### PR TITLE
More complete Result

### PR DESCRIPTION
**Pull request by psznm**
Hello, I may went a little overboard with this contribution. 

I had some conflicts of functionality/naming and considered deprecating original code, but the naming conflict was not really solvable (can't come up with any other good name than AsyncResult) so I decided to go for full replacement instead of deprecation. These conflicts are:

1. `.isError` -> `.isErr()` because I was not able to make type narrowing work with simple property.
2. type `AsyncResult` -> class `AsyncResult` because just wrapping `Result` with Promise is not good enough.
3. removed `toResultAsync` in favor of new `AsyncResult.fromPromise`

For this reason this PR aims for new major version of this package.

This PR adds many functionalities inspired by Rust Result, and also adds whole new class to make dealing with async Results a pleasure.

I am looking forward to your feedback.

### CODE REVIEW

File: src/asyncResult.test.ts [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+        }).promise;
+    });
+
+    it('inspectErr gets called with Result error if Result is Err', async () => {
+        let inspected: number|undefined = undefined;
+        await AsyncResult.fromResult(Err(5)).inspectErr(val => inspected = val).promise;
+        expect(inspected).toEqual(5);
+    });
+
+    it('inspectErr does not get called if Result is Err', async () => {
```
Comment (Kostayne) [inline src/asyncResult.test.ts]: I guess it should be "if Result is Ok" here

File: src/asyncResult.test.ts [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+    });
+    it('async mapOrElse runs provided fallback on err of Result, and creates new Ok Result', async () => {
+        const res1 = AsyncResult.fromResult(Err(5))
+        const res2 = res1.mapOrElse(() => { 
+            expect('to not have been called').toBe(true)
+            return 0;
+        }, err => err+1);
+        expect(await res2.unwrap()).toEqual(6)
+    });
+});
```
Comment (Kostayne) [inline src/asyncResult.test.ts]: Same tests with different names, line 173 requires `async` keyword before callback

File: src/result.ts [hunk 3]
```diff
@@ -115,19 +365,21 @@ export function toResult<T, E>(fn: () => T): Result<T, E> {
...
+export function resultify<T, E, F extends Fn<T>>(fn: F, mapErr: (err: unknown) => E): ResFn<T, E, F> {
+    return (...params) => { 
+        try {
+            const val = fn(...params);
+            return Ok(val);
+        } catch(e) {
+            return Err(mapErr(e));
+        }
     }
-
```
Comment (Kostayne) [inline src/result.ts]: The current solution has drawbacks:
- Result value type is not picked up from function type, so we get Result<unknown, unknown> by default
- Error mapper is required, but it's not always the case for the user
- Error passed in mapper is always unknown, there is no way to specify a type

We can solve these problems with this code:
```ts
type ResFn<E, F extends Fn<ReturnType<F>>> = (...params: Parameters<F>) => Result<ReturnType<F>, E>;

/**
 * @description Catches exceptions and converts them into Result
 * @example
 * const rawFn = (a: number) => throw new Error('not today');
 * const fn = resultify(rawFn, err => err.msg);
 * const res = fn(2);
 * res.err() // 'not today'
*/
export function resultify<F extends Fn<ReturnType<F>>, E = Error, E2 = Error>(fn: F, mapErr: (err: E) => E2): ResFn<E2, F>;
/**
 * @description Catches exceptions and converts them into Result
 * @example
 * const rawFn = (a: number) => a;
 * const fn = resultify(rawFn);
 * const res = fn(2);
 * res.ok() // 2
*/
export function resultify<F extends Fn<ReturnType<F>>, E = Error>(fn: F): ResFn<E, F>;

export function resultify<F extends Fn<ReturnType<F>>, E = Error>(fn: F, mapErr = (e: E) => e) {
    return (...params: any[]) => { 
        try {
            const val = fn(...params);
            return Ok(val);
        } catch(e) {
            return Err(mapErr(e as E));
        }
    }
}
```

I added function overloading to enforce the same error type as on input if there is no mapper provided.

File: src/asyncResult.ts [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+type AsyncMapped<T> = T|Promise<T>
+
+type AsyncFn<T> = (...params: any[]) => Promise<T>;
+type AsyncResFn<T, E, F extends AsyncFn<T>> = (...params: Parameters<F>) => AsyncResult<T, E>;
+/**
+ * @description Creates function returning AsyncResult from provided async function and error mapper
+ */
+export function asyncResultify<T, E, F extends AsyncFn<T>>(fn: F, mapErr: (err: unknown) => E): AsyncResFn<T, E, F> {
+    return (...params) => AsyncResult.fromPromise(fn(...params)).mapErr(mapErr);
+}
```
Comment (Kostayne) [inline src/asyncResult.ts]: It would be great if asyncResultify were written as a non-asynchronous version (see the previous comment)

File: src/asyncResult.ts [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+export class AsyncResult<Val, Err> {
+
+    /**
+     * @param promise
+     * @description AsyncResult.fromPromise(promise, mapError) will return AsyncResult containing the promise.
+     * @example
+     * const resPromise = AsyncResult.fromPromise(Promise.resolve(1));
+     * // Type of resPromise is AsyncResult<number, unknown>
+     */
+    static fromPromise<Val>( promise: Promise<Val>): AsyncResult<Val, unknown> {
```
Comment (Kostayne) [inline src/asyncResult.ts]: Current
```ts
const asyncRes = AsyncResult.fromPromise(Promise.resolve(1));
```

Shorter alternative that nice to have
```ts
const asyncRes = toAsyncResult(Promise.resolve(1));

/**
 * @param promise 
 * @description Converts promise resolve / reject to Ok() or Err()
 * @example
 * const promise = Promise.reject();
 * const result = toAsyncResult(promise);
 */
export function toAsyncResult<T, E = Error>(promise: Promise<T>): AsyncResult<T, E> {
    return AsyncResult.fromPromise<T, E>(promise);
}
```

File: README.md [hunk 6]
```diff
@@ -57,10 +112,11 @@ function Err<Err>(err: Err): ERR<Err>;
 // Convert exceptions into errors & function result into Ok
 function toResult<Val, ERR>(fn: () => Val): Result<Val, ERR>;
 
-// Convert promise reject into errors & promise resolve into Ok
-async function toResultAsync<Val, ERR>(fn: () => Promise<Val>): Promise<Result<Val, ERR>>;
+// Convert function returning T and error mapper returning E to new function returning Result<T, E>
+function resultify<T, E, F extends Fn<T>>(fn: F, mapErr: (err: unknown) => E): ResFn<T, E, F>;
 
-type AsyncResult<Val, ERR> = Promise<Result<Val, ERR>>;
```
Comment (Kostayne) [inline README.md:63]: Typing `Promise<Result<Val, Err>>` was not cool, too many generic layers. We can't return `AsyncResult` instance directly from async function. So the type is still needed. Just change it to `AsyncResultT`.

```ts
type AsyncResultT<Val, Err = Error> = Promise<AsyncResult<Val, Err>>;
```

File: README.md [hunk 7]
```diff
@@ -26,21 +26,76 @@ pnpm add base-ts-result
...
+Class that contains operation result and interaction methods for async code
+``` ts
+type AsyncMapped<T> = T|Promise<T>
+// Async result implementation for more ergonomic usage of Results with async code
+class AsyncResult<Val, Err> {
+    // Contained promise
+    promise: Promise<Result<Val, Err>>; 
+
+    // Creators
+    static fromPromise<Val>(promise: Promise<Val>): AsyncResult<Val, unknown>;
```
Comment (Kostayne) [inline README.md:70]: Can't specify error type here, a better alternative:
```ts
static fromPromise<Val, Err = Error>(promise: Promise<Val>): AsyncResult<Val, Err> {
```

File: src/asyncResult.test.ts (hunk 1 above)
Comment (psznm) [reply src/asyncResult.test.ts]: Fixed

File: src/asyncResult.test.ts (hunk 2 above)
Comment (psznm) [reply src/asyncResult.test.ts]: fixed

File: src/result.ts (hunk 3 above)
Comment (psznm) [reply src/result.ts]: I have fixed the first and second points. I would like to discuss the third one with you.

The error passed to mapErr directly translates to the error caught by resultify. That error can safely only be typed unknown, or unsafely any.

See this example
``` ts
const test = () => {
    try {
        throw 1;
    } catch (err: number) { // Catch clause variable type annotation must be 'any' or 'unknown' if specified. [TS1196]
        return err.message;
    }
}
```
If the error would be any, user could specify type, but it would be unsafe. Like this user has to narrow his error type to ensure typesafe E of his Results

File: README.md (hunk 6 above)
Comment (psznm) [reply README.md:63]: I would like to not override meaning of basically same name to two quite different things. How does "ResultPromise" for `Promise<Result<Val, Err>>` and "AsyncResultPromise" `Promise<AsyncResult<Val, Err>>` sound to you?

File: README.md (hunk 7 above)
Comment (psznm) [reply README.md:70]: Please, see my comment about type safety. I do not think user should be able to just specify error type, but should type safely narrow type of his error instead.

I was thinking of adding mapErr parameter to this function for convenience, but I decided against it because `AsyncResult.fromPromise(promise, () => 'err')` brings very little convenience over `AsyncResult.fromPromise(promise).mapErr(() => 'err')`, which offers more minimal API and IMO more clarity.

File: src/result.ts (hunk 3 above)
Comment (Kostayne) [reply src/result.ts]: I have thought about that and came to a solution that satisfies both the convenience and safety. We can reduce all error types to one.

The result code will look like this:
```ts
export function resultify<F extends Fn<ReturnType<F>>>(fn: F, mapErr = (e: BaseError) => e) {
    return (...params: any[]) => { 
        try {
            const val = fn(...params);
            return Ok(val);
        } catch(e) {
            const baseErr = new BaseError(e, '');
            return Err(mapErr(baseErr));
        }
    }
}

/**
 * @description converts non Error values to Error
 */
export class BaseError extends Error {
    /**
     * @description exotic value that was thrown
     */
    public origValue: unknown;

    name = "BaseError";

    // specified any to not write additional casts
    constructor(origValue: any, msg: string) {
        super(msg);

        const isErr = origValue instanceof Error;

        if (!isErr) {
            this.origValue = origValue;
        }

        // setting origVal.toString() into message
        if (typeof origValue?.toString == 'function' && !isErr) {
            const newMsg = origValue.toString();

            if (typeof newMsg === 'string') {
                const type = Array.isArray(origValue) ? 'array' : typeof origValue;
                this.message = `Caught exotic value (${type}): ${newMsg}`;
            }
        }

        // copy orig error params
        if (isErr) {
            this.stack = origValue.stack;
            this.message = origValue.message;
            this.name = origValue.name;
        }
    }
}
```

File: README.md (hunk 6 above)
Comment (Kostayne) [reply README.md:63]: I think `ResultPromise` and `AsyncResultPromise` are great

File: README.md (hunk 7 above)
Comment (psznm) [reply README.md:70]: fromPromise now returns `AsyncResult<Val, ResultCaughtError>`

File: README.md (hunk 6 above)
Comment (psznm) [reply README.md:63]: Added

File: src/asyncResult.ts (hunk 5 above)
Comment (psznm) [reply src/asyncResult.ts]: Added

File: src/asyncResult.ts (hunk 4 above)
Comment (psznm) [reply src/asyncResult.ts]: Done.

File: src/result.ts (hunk 3 above)
Comment (psznm) [reply src/result.ts]: I really like the idea, but there is one issue why we shouldn't do it exactly like this. For example `exec` from `child_process` rejects with `ExecException` which is instance of `Error` but also has properties `killed`, `cmd`, `code`, etc... These additional properties would be lost here. Also we would have lost the ability to narrow type to `ExecException` because it does not extend our `BaseError`

So I thought maybe we could just passthrough thrown Errors, and wrap only non-Error thrown values. But then accessing the original thrown value would be quite awkward. For example
```
if ('origValue' in err) {
    console.log(err.origValue);
}
```

But maybe that is not really an issue...? Because honestly, how often does someone throw non-Error anyway, right...? So that is what I went with for now.

Comment (Kostayne) [reply src/result.ts]: You can easily throw a string, that's a pretty common case, so we can't ignore it. And yep, we still can provide a better solution. We will check & change prototype if needed, also copy the original properties.  

Js is one of the most powerful languages, there is nothing we can't do. 

The code:
```ts
import clone from "clone";

/**
 * @description converts non Error values to Error
 */
export class BaseError extends Error {
    /**
     * @description exotic value that was thrown
     */
    public origValue?: unknown;

    name = "BaseError";

    // specified any to not write additional casts
    constructor(errVal: any, msg: string) {
        super(msg);

        const isErr = errVal instanceof Error;

        if (!isErr) {
            this.origValue = errVal;
        }

        // setting origVal.toString() into message
        if (typeof errVal?.toString == 'function' && !isErr) {
            const newMsg = errVal.toString();

            if (typeof newMsg === 'string') {
                const type = Array.isArray(errVal) ? 'array' : typeof errVal;
                this.message = `Caught exotic value (${type}): ${newMsg}`;
            }
        }

        // copy orig error params
        if (isErr) {
            const errProto = Object.getPrototypeOf(errVal);

            // Copy prototype and properties if errVal is
            // a subclass of Error
            if (errProto !== Error.prototype) {
                Object.setPrototypeOf(this, errProto);

                const copy = clone(errVal);
                Object.assign(this, copy);
            }

            this.stack = errVal.stack;
            this.message = errVal.message;
            this.name = errVal.name;
        }
    }
}
```

Comment (psznm) [reply src/result.ts]: I think this is not a good solution.

1. We are forcing everyone to build stack trace, any time they throw non-Error. Building stack trace can be very expensive.
2. We are also building second stack trace when Error is thrown only to discard it.
3. To know whether "exotic value" was thrown, user still has to check for `'origValue' in err`, so no improvement over current code. User could now also check if `origValue` is undefined, but that is less correct, since `undefined` could be thrown and assigned to `origValue`.
5. User has no way to know that their error typed `BaseError` can be `instanceof` any other `Error` based class. Unlike `unknown` or `Error`, `BaseError` does not feel like something you can narrow down to another Error based class from different library.
6. People will expect `BaseError` in their err, and suddenly see in their log `ExecException`. That seems to be very high surprise factor.
7. I do not think there is any real benefit over current code.

I think that err in case of (async)resultify and fromPromise should just remain unknown. We should let user handle their errors, and not do anything obscure and unexpected.  I would accept wrapping non-Error values because I think that throwing non-Errors is very niche, but even that feels wrong because of point 1. If you are worried about convenience, know that both resultify and fromPromise are intended only to aid in starting using of Results in existing codebases without having to rewrite existing functions for which one would want to use Results. New functions should already be returning Result/AsyncResult where both T and E are already correctly typed. Also writing custom helper to wrap error in whatever container is trivial and in my opinion should not be part of this library as it takes control/freedom away from the user of this library.

TLDR; I would prefer reverting `err` back to `unknown` for reasons above, but would also accept only wrapping non-Errors.
Please, let me know about your opinion on the subject.

Comment (Kostayne) [reply src/result.ts]: The code above allowed type narrowing with 'instanceof', but, never mind, the inheritance from Error introduces too many issues with the stack trace. You are absolutely right about just passing the original error instance; there is no point in modifying it.

I think we have a better way to handle exotic error values. Simply create a plain object that implements the Error interface instead of creating an Error instance. This approach eliminates the computation of a stack trace pointing to the library code and improves performance.

I don't see any serious problems in this simpler implementation.
```ts
type BaseError = Error & {
    /**
     * @description exotic value that was thrown
     */
    origValue?: unknown;
}

const toBaseError = (errVal: any): BaseError => {
    const isErr = errVal instanceof Error;

    if (isErr) {
        // regular error, just pass it throw
        return errVal as BaseError;
    }

    // Error value is not an error instance
    const resErr = {} as BaseError;

    resErr.origValue = errVal;
    resErr.name = 'BaseError';

    // setting origVal.toString() into message
    if (typeof errVal?.toString == 'function') {
        const newMsg = errVal.toString();

        if (typeof newMsg === 'string') {
            const type = Array.isArray(errVal) ? 'array' : typeof errVal;
            resErr.message = `Caught a plain value (${type}): ${newMsg}`;
        }
    }

    return resErr;
};
```

Comment (psznm) [reply src/result.ts]: I am happy to hear that we are on the same page now. My implementation differs very slightly from yours, but functionally it should be the same (EDIT: I made the wrapper object be a class to be able to specify .toString() for nice implicit stringification).  I think this should resolve the last issue of this PR?

Comment (Kostayne) [reply src/result.ts]: All great, only one detail left. 

The `Type ResultBaseError = BaseError | Error;` will not show the existence of the 'origValue' property. The only way for a user to get know about it - go to the function definition.

This will fix the problem:
```ts
export type ResultBaseError = Error & {
    origValue?: unknown;
};
```

Comment (psznm) [reply src/result.ts]: Changed
